sound/soc/sof/intel/hda-dsp.c

```diff
@@ -887,10 +887,63 @@ int hda_dsp_suspend(struct snd_sof_dev *sdev, u32 target_state)
 	return snd_sof_dsp_set_power_state(sdev, &target_dsp_state);
 }
 
+static unsigned int hda_dsp_check_for_dma_streams(struct snd_sof_dev *sdev)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct hdac_stream *s;
+	int active_streams = 0;
+	int sd_offset;
+	u32 val;
+
+	list_for_each_entry(s, &bus->stream_list, list) {
+		sd_offset = SOF_STREAM_SD_OFFSET(s);
+		val = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
+				       sd_offset);
+		if (val & SOF_HDA_SD_CTL_DMA_START)
+			active_streams |= BIT(s->index);
+	}
+
+	return active_streams;
+}
+
+static int hda_dsp_s5_quirk(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	ret = hda_dsp_ctrl_link_reset(sdev, false);
```

Comment:
I might be too cautious, but I would add `usleep_range(500, 1000);` before taking the HDA out of reset, I'm not sure how long the `hda_codec_i915_display_power()` takes. Yous to be on the safe side?

Reply:
True @ujfalusi . At least i915 call is not deterministic in timing, so better not to rely on it. Updated.